src/cluster.c

```diff
@@ -4647,31 +4838,56 @@ void clusterUpdateSlots(client *c, unsigned char *slots, int del) {
     }
 }
 
-void addNodeReplyForClusterSlot(client *c, clusterNode *node, int start_slot, int end_slot) {
-    int i, nested_elements = 3; /* slots (2) + master addr (1) */
-    void *nested_replylen = addReplyDeferredLen(c);
-    addReplyLongLong(c, start_slot);
-    addReplyLongLong(c, end_slot);
-    addReplyArrayLen(c, 3);
-    addReplyBulkCString(c, node->ip);
+void addNodeToNodeReply(client *c, clusterNode *node) {
+    addReplyArrayLen(c, 4);
+    if (server.cluster_preferred_endpoint_type == CLUSTER_ENDPOINT_TYPE_IP) {
+        addReplyBulkCString(c, node->ip);
+    } else if (server.cluster_preferred_endpoint_type == CLUSTER_ENDPOINT_TYPE_HOSTNAME) {
+        addReplyBulkCString(c, node->hostname ? node->hostname : "?");
+    } else if (server.cluster_preferred_endpoint_type == CLUSTER_ENDPOINT_TYPE_UNKNOWN_ENDPOINT) {
+        addReplyNull(c);
+    } else {
+        serverPanic("Unrecognized preferred endpoint type");
+    }
+    
     /* Report non-TLS ports to non-TLS client in TLS cluster if available. */
     int use_pport = (server.tls_cluster &&
                      c->conn && connGetType(c->conn) != CONN_TYPE_TLS);
     addReplyLongLong(c, use_pport && node->pport ? node->pport : node->port);
     addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
 
+    /* Add the additional endpoint information, this is all the known networking information
+     * that is not the preferred endpoint. */
+    void *deflen = addReplyDeferredLen(c);
+    int length = 0;
+    if (server.cluster_preferred_endpoint_type != CLUSTER_ENDPOINT_TYPE_IP) {
+        addReplyBulkCString(c, "ip");
+        addReplyBulkCString(c, node->ip);
+        length++;
+    }
+    if (server.cluster_preferred_endpoint_type != CLUSTER_ENDPOINT_TYPE_HOSTNAME
+        && node->hostname)
+    {
+        addReplyBulkCString(c, "hostname");
+        addReplyBulkCString(c, node->hostname);
+        length++;
+    }
+    setDeferredMapLen(c, deflen, length);
```

Comment:
So the 4th element "additional networking information" is a map with info-type as the keys? A map can't have duplicate keys, or can it? For multiple IP addresses, duplicate keys are needed.

Reply:
I'm still of the opinion that we don't want to support multiple endpoints of the same type, so I think a map is fine.

Reply:
@madolson we do have a conflict here:
* No multiple endpoints of the same type - I agree in principle, we'll just get clients confused.
* No specific ipv6/ipv4 types - I agree with that as well, it's not something we currently do.
But if I'm forced to choose, I'd give up the first to support IPv4 and IPv6 concurrently, simply because I don't see the real downside of the response format allowing that. Not feeling too strong about it though.

Reply:
We could also re-evaluate the possibility of not tagging them with a type and let the client figure out what they are. Then, we IP versions and multiple of the same type become non-issues.

Reply:
@yossigo I don't think the resolution to supporting both ipv4 and ipv6 is to show both of them with the same name and have the clients check all the entries and figure out what they want to do. We only gossip a single IP address on the clusterbus today (which can be IPv6). If we want to introduce more, we would need to a way to name them/announce them, so I think however we resolve that naming would give us a new name to add to this map.

I also will probably advocate that if we go down the path of supporting dual stack IPv4 + IPv6 configurations, we just determine which the cluster responds with the same way we do with TLS. If a client connects over IPv4, we respond with IPv4, and the same with IPv6.

Reply:
@madolson I just want to point out that as the interface design here involves clients, it is likely to outlive many of the current server-side implementation details. You're basically saying any future extension will need its own new and unique identifier. I'm OK with that.